buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java

```diff
@@ -0,0 +1,187 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle;
+
+import groovy.lang.Closure;
+import org.gradle.api.Action;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.api.artifacts.dsl.DependencyHandler;
+import org.gradle.api.file.CopySpec;
+import org.gradle.api.file.FileCopyDetails;
+import org.gradle.api.file.FileTree;
+import org.gradle.api.file.RelativePath;
+import org.gradle.api.plugins.ExtraPropertiesExtension;
+import org.gradle.api.tasks.Copy;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JdkDownloadPlugin implements Plugin<Project> {
+
+    private static final Pattern JDK_VERSION = Pattern.compile("(\\d+)(\\.\\d+\\.\\d+)?\\+(\\d+)");
+
+    private List<String> ALLOWED_PLATFORMS = Collections.unmodifiableList(Arrays.asList("linux", "windows", "darwin"));
+
+    @Override
+    public void apply(Project project) {
+
+        // create usesJdk task "method"
+        project.getTasks().all(task ->
+            task.getExtensions().findByType(ExtraPropertiesExtension.class).set("usesJdk", new Closure<Void>(project, task) {
+                public void doCall(String version, String platform) {
+                    if (version == null) {
+                        throw new IllegalArgumentException("version must be specified to usesJdk");
+                    }
+                    if (platform == null) {
+                        throw new IllegalArgumentException("platform must be specified to usesJdk");
+                    }
+                    if (ALLOWED_PLATFORMS.contains(platform) == false) {
+                        throw new IllegalArgumentException("platform must be one of " + ALLOWED_PLATFORMS);
+                    }
+                    ExtraPropertiesExtension extraProperties = task.getExtensions().findByType(ExtraPropertiesExtension.class);
+                    if (extraProperties.has("jdkHome")) {
+                        throw new IllegalArgumentException("jdk version already set for task");
+                    }
+
+                    // ensure a root level jdk download task exists
+                    setupRootJdkDownload(project.getRootProject(), platform, version);
```

Comment:
Could we set up a single task to download all as we do for test clusters rather than one for  each version ? Having fewer tasks will make configuration faster and I find it easier to follow what's going on. 
It could also be a lazily created task that's always set up and  then have the tasks that require it depend on it

Reply:
I might be missing it, but is `task` made dependent on this ?

Reply:
I do not think we should have a single configuration for all jdks we download. That means resolving the configuration triggers download of all of them. Additionally, the way this works is to be able to resolve a configuration to a single file.

> is task made dependent on this ?

The root project does the extraction, and makes the extracted dir available as an artifact of the root project. The tasks in subprojects add a normal dependency on that "artifact", through a separate configuration in the subproject, and the task then depends on that local subproject configuration.

Reply:
I wasn't referring to having a single configuration, we actually need those to be separate to be able to have different versions, but we can extract all in the same task, having multiple tasks for it doesn't buy us anything.

Reply:
Having a single task would mean the task depends on all those configurations, thus still resolving all the artifacts at the same time. We need a 1-1 mapping to ensure one extraction is done.

Reply:
Is that all that bad?   The configurations are cached, and the task has up to date checks .  For most things this all the JDKs will be needed anyway due to the number of tests that depend on the JDK.

Reply:
> For most things this all the JDKs will be needed anyway

I don't think that is true. When I run a packaging test for a single distro, I need one jdk, not many, and certainly not the ones from non linux.